value/valuereflect.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package value
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+var reflectPool = sync.Pool{
+	New: func() interface{} {
+		return &valueReflect{}
+	},
+}
+
+// NewValueReflect creates a Value backed by an "interface{}" type,
+// typically an structured object in Kubernetes world that is uses reflection to expose.
+// The provided "interface{}" may contain structs and types that are converted to Values
+// by the jsonMarshaler interface.
+func NewValueReflect(value interface{}) (Value, error) {
+	if value == nil {
+		return NewValueInterface(nil), nil
+	}
+	return wrapValueReflect(reflect.ValueOf(value))
+}
+
+func wrapValueReflect(value reflect.Value) (Value, error) {
+	// TODO: conversion of json.Marshaller interface types is expensive. This can be mostly optimized away by
+	// introducing conversion functions that do not require going through JSON and using those here.
+	if marshaler, ok := getMarshaler(value); ok {
+		return toUnstructured(marshaler, value)
```

Comment:
Will this actually be used often?

Reply:
Yeah, Time and Quantity are common enough that this gets called quite a bit. I have a follow up PR to optimize those conversions more since round tripping through json is expensive.